Unify gradient accumulation between distributed autograd and local autograd
engine.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33427 Avoid clone for sparse tensors during accumulation of grads.
* **#33214 Unify gradient accumulation between distributed autograd and local autograd
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

pritamdamania committed:
Unify gradient accumulation between distributed autograd and local autograd
engine.

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

[ghstack-poisoned]

pritamdamania committed:
Update on "[NOT READY FOR REVIEW] Unify gradient accumulation between distributed autograd and local autogradengine."

engine.**

engine.

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot retest this please

pritamdamania committed:
Update on "[NOT READY FOR REVIEW] Unify gradient accumulation between distributed autograd and local autograd engine."


Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

[ghstack-poisoned]

pritamdamania committed:
Update on "[NOT READY FOR REVIEW] Unify gradient accumulation between distributed autograd and local autograd engine."


Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

[ghstack-poisoned]

albanD reviewed (commented):
Looks mostly good.
Some cpp question about simplifying the code.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
This wouldn't work as expected since grads[i] is const (since `const variable_list& grads`) and would anyways incur a copy as a result. The reason grads is const is because the Future holds a reference to the value and only allows callbacks to operate on a const ref of the value.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
local_grads stores the result from the first iteration for `ExecMode.LOCAL` and this value is then used in later iterations to validate that local execution and remote execution produce the same grads (see _verify_backwards)

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd engine."


Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."


engine.

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."

engine.**

engine.

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]

ezyang commented:
Alban requested that I review C++ move semantics. I am reviewing this aspect of the PR only.

ezyang commented on torch/csrc/autograd/engine.cpp:
I don't think the log line here is appropriate.

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
A little confused why you don't `std::move(new_grad)` here

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
Obviously the `std::move` does nothing if `accumulateGradAndCallHooks` takes by reference, but I don't see any good reason for it to take the arg by reference.

pritamdamania87 commented on torch/csrc/autograd/engine.cpp:
Do you feel we shouldn't log this information (because it could spew a lot of logs) or is there actually something wrong with the log message itself?

pritamdamania87 commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
The reason `accumulateGradAndCallHooks` takes a const reference is because we would like to avoid a copy that would increment the Tensor's refcount. We can use std::move() here to avoid the refcount bump, although we can't use std::move() when we call `accumulateGradAndCallHooks` since the grad there is a const reference that is owned by the Future and the Future only allows callbacks to operate on a const ref of the value (which makes sense for the Future API).

ezyang commented on torch/csrc/autograd/engine.cpp:
Mostly spew. The log line doesn't seem actionable.

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
Let me put it differently. Why do you move the tensor out of the passed in `variables_list`? If you're going to pass the tensor on by reference (as happens here), then there's no point in the move. Basically, the code is inconsistent: if it moves, it should move all the way to a point where the moves pay off (called function takes ownership); otherwise, you should just pass by reference all around. It sounds like you're in the latter situation.

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."

engine.**
engine.**
engine.**
engine.**
engine.**
engine.**

engine.

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]

pritamdamania87 commented:
@ezyang Could you take another look at this PR? I've addressed your comments regarding std::move and the log spew.

albanD commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
Ho right, this is special logic in the distributed tests. Looks good.

albanD commented on torch/csrc/distributed/autograd/context/context.cpp:
You can never have `post_hooks` in the distributed engine?

albanD commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
nit: maybe this does not need to be a special parameter. We just want to add `1` to `num_expected_refs` right?

pritamdamania87 commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
Good point, let me fix this.

albanD commented on torch/csrc/distributed/autograd/context/context.cpp:
Sounds good. Do you have an assert to make sure this does not happen for now?

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."

engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**

engine.

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
nit: accumulates grad use the given function.

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
`update_grad_fn` is not listed here and `has_post_hooks` should be removed.

albanD commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
Why is this one not using the `update_grad_fn`? That would prevent the gradient to be recorded by the distributed autograd no?

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
nit: `update_grad_fn` can be misleading in the autograd engine as `.grad_fn` is an important field on Tensors. And this function has nothing to do with that field.

albanD commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
I'm curious why this is not detected by the tests? I did not went through all the tests in details.

albanD reviewed (changes requested):
Looks mostly good, but some suspicious bits.

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."

engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**

engine.

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
This is very surprising.
Does that mean that this function will either modify `variable_grad` inplace or call `update_grad` ?
This should be explicitly documented at least.

But it sounds like a quite weird thing to do.
Especially given that if you have get a reference to the variable_grad, you can just set it to modify the original Tensor inplace (coming either from variable.grad() or accumulatedGrads_[idx]).
That would make the API clearer no?

ezyang commented on torch/csrc/autograd/engine.cpp:
It would be good to specify in the docs of `captured_vars_` that it is moved out of (and thus no longer valid on the graph task) after post processing.

ezyang commented on torch/csrc/distributed/autograd/context/context.h:
I was taking a second look and I was getting bad vibes about the recursive mutex, so I took a closer look here. I'm not convinced by this explanation to Alban's comment. A few things:

1. The problem with recursive mutexes is a lot of things just don't "work correctly" when you do a recursive mutex. For example, if you do a condition variable wait on a mutex that has been recursively locked, you will still maintain the lock in the outer scope (likely leading to a deadlock). I don't think the code as you've written here commits this mistake, but there are just a lot more things that can go wrong when a recursive mutex is involved and so it just makes it harder to code review the code in the future.

2. The most common reason to introduce a recursive mutex is because you can't conveniently rearrange your code so that you never call locking functions from inside locking functions; e.g., maybe there is a pretty nested call chain and you don't want to created an unlocked version of every function. But that's not the case here: you're directly calling `set_exception` which is the only thing which recursively takes out the lock; you then observe that it might call `resetGraphTask` eventually but we see:

```
void GraphTask::set_exception(
    std::exception& e,
    const std::shared_ptr<Node>& fn) {
  std::unique_lock<std::mutex> lock(mutex_);
  if (!has_error_.load()) {
    if (AnomalyMode::is_enabled() && fn) {
      fn->metadata()->print_stack();
    }
    has_error_ = true;
    // Careful: setting the future_result_ can trigger DistAutogradContext to
    // resetGraphTask(), sometimes deleting this underlying GraphTask.
    // Don't touch *this after setError() below, and release the lock early, to
    // avoid unlocking this->mutex_ after setting the future.
    std::shared_ptr<FutureVariableList> future_result = future_result_;
    lock.unlock();
    if (!future_result->completed()) {
      future_result->setError(e.what());
    } else {
      TORCH_INTERNAL_ASSERT(future_result->hasError());
    }
  }
}
```

This avoids the recursive mutex call, AND it is good logic: at the point in time you are calling setError on the future result, you don't need the lock on the GraphTask, so you ought to release it before calling arbitrary user code.

More concretely, I don't see why you can't refactor `set_exception` into a version that takes in the lock that it should release before calling the future callback, and then call that direct version with the parent lock. Then, as far as I can tell, you don't need a recursive mutex at all. This is a much more local change.

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.h:
`std::function` here looks slow, because we have to convert the lambda object into the `std::function` representation. Example https://github.com/facebook/folly/issues/344

It seems to me that you just wanted to reuse code here, and effectively just inline `accumulateGradAndCallHooks` into the call sites. You can do that if you convert this into a template function, templated over the (unnamed) lambda type, then you can have your cake and eat it too.

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
If you're going to take `grad_update` by rvalue reference, you should cash in on it and `grad = std::move(grad_update)`

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
??? This doesn't make any sense to me. All methods on Tensor should be const-qualified (even if they mutate the tensor--constness is a lie). Why do you need to work with a non-const tensor?

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
So, then you do `auto new_grad = std::move(grads[0])`

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
I have rereviewed the code after your changes and I think the original move was justified. See my comments above. You should put back the move here.

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
You've completely lost the comment about why `!post_hooks().empty()` is used here

ezyang reviewed (changes requested):
Apologies about the late review; apparently, it's not possible to do a "C++ only code review", I actually have to understand what the code is doing. 8) I think some of the changes pessimize code for no good reason, and we should avoid this before landing. I haven't reviewed the new call site from distributed autograd in detail yet.

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."

engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**
engine.**

engine.

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]

pritamdamania87 commented:
@ezyang I've refactored the code a bit to get rid of the recursive_mutex, could you take another look?

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
Are you consuming the Future when you access the grad? In that case you can add a `&&` overload so you can move the Tensor result out.

pritamdamania87 commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
Well the grad is accessed inside the callback of the future: https://github.com/pytorch/pytorch/blob/dbe850af5b941c6f0ae1ed50e7d6b779db55f5af/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L216. I don't think it's a good idea to move out the Future's value at that point since in general you don't know if there might be other callbacks running on the Future after that one.

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
Well, the other trick when you (1) take ownership of the incoming tensor unconditionally and (2) need both the const ref and rvalue overloads, is to take the input tensor by value, e.g., `Tensor input`. When it comes in as a const ref that will trigger a ref count bump; when it comes in as an rvalue, you'll move construct and no refcount bump occurs.

gqchen commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
The additional cost for using const ref is we need to make a shallow copy inside accumulateGradAndCallHooks.  Given the current code structure, this seems to be reasonable to me. In the future, when we move dist grad accumulation to be inline (just like local grad), we can change this to std::move.

gqchen reviewed (commented):
This PR looks good to me.

albanD commented on torch/csrc/autograd/engine.cpp:
What is the difference with `future->setError()` ?

albanD commented on torch/csrc/distributed/autograd/context/context.cpp:
That doesn't look like the right place to enforce this?
Is this because the AccumulateGrad is not a Node here?

albanD commented on torch/csrc/distributed/autograd/context/context.cpp:
The whole point of breaking `set_exception` in two is to be able to release this lock in the middle?
Why is it not enough to make a local copy in this scope of `graphTask_` to make sure it does not get deleted too early?

albanD reviewed (commented):
Looks mostly good, but the lock and lifetime logic seems quite complex and I am not convinced it needs to be.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/context.cpp:
Making the local copy would work, although its nicer to avoid holding a lock when we mark the future as completed. Since the future, would trigger callbacks and its better not to hold locks there. You can see the discussion on https://github.com/pytorch/pytorch/pull/33156.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/context.cpp:
The other reason for breaking up set_exception is to avoid the mutex. If we don't unlock here, the callbacks called from future->setError will try to lock the mutex again.

albanD commented on torch/csrc/distributed/autograd/context/context.cpp:
I guess if there is no plan in supporting `create_graph=True`, that is ok.

albanD commented on torch/csrc/autograd/engine.cpp:
So it does exactly the same as:
```cpp
    if (!future_result->completed()) {
      future_result->setError(e.what());
    } else {
      TORCH_INTERNAL_ASSERT(future_result->hasError());
    }
```
Or is there any difference?

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."


Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."

engine.**

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]

pritamdamania87 commented:
Update the PR with the following changes:

1. Added an ASSERT to ensure the Future that holds onto the gradient is cleaned up before we're done with the backward pass to ensure we don't leak references to the gradients elsewhere.
2. Removed const ref for `accumulateGradAndCallHooks`, used std::move() where possible and incremented `num_expected_refs` in cases where we couldn't use std::move() (due to const ref)
3. Passed in a copy of GraphTask to set_exception instead of a const ref to get rid of the constraint of not touching *this in that method.

hczhu commented on torch/csrc/autograd/functions/accumulate_grad.h:
Just being curious. what happens to ` += ` if `grad_variable` is sparse, and `new_grad` is dense?

hczhu commented on torch/csrc/autograd/functions/accumulate_grad.h:
Do we want to spell `T` out to avoid this function template and move the impl to .cpp file?

```
...
std::function<void(at::Tensor&& grad_update)>&& update_grad) {
```

hczhu commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
I'd suggest leveraging `unsafeGetTensorImpl()` and a hash table to avoid the quadratic time complexity here.

```
std::unordered_map<const TensorImpl*, int> tensor_to_refs;
for (size_t i = 0; i < grads.size(); i++) {
  const auto num_expected_refs = (tensor_to_refs[grads[i].unsafeGetTensorImpl()] += 1);

```

Note that `is_same()` is implemented as.

https://github.com/pytorch/pytorch/blob/f9f135c5d8fc1986e112e62473b3498cdf4d1c43/aten/src/ATen/templates/TensorBody.h#L174-L176

hczhu commented on torch/csrc/utils/future.h:
The comment seems not right. `lock` is holding the mutex.

hczhu commented on torch/csrc/utils/future.h:
nit:

```
std::vector<Callback> cbs(std::move(callbacks_);
```

ezyang commented:
> Update the PR with the following changes:

Thank you for the description of changes, I appreciate it.

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.h:
!!! Keep this one a reference! You don't move into this one, it's not taken ownership of in the body here, should be a `const Tensor& variable_grad` !!!

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.h:
It doesn't work:

```
>>> i = torch.LongTensor([[0, 1, 1],
...                           [2, 0, 2]])
>>> v = torch.FloatTensor([3, 4, 5])
>>> s = torch.sparse.FloatTensor(i, v, torch.Size([2,3]))
>>> d = s.to_dense()
>>> s += d
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: add(sparse, dense) is not supported. Use add(dense, sparse) instead.
```

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.h:
I would like to not have to make comments like this in the future :>

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.h:
If you spell it as `std::function` we have to do an expensive std::function construction every time you want to call this function. Templating lets the compiler inline the lambda so no construction happens.

ezyang commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
@pritamdamania87  and I discussed this and we're just going to delete this "optimization" as it is incredibly unlikely to happen in practice

ezyang reviewed (approved):
Approving conditional two things

albanD commented on torch/csrc/autograd/engine.cpp:
Why do you remove the reference here?

pritamdamania87 commented on torch/csrc/autograd/engine.cpp:
The reason is that with the const ref, the GraphTask can be deallocated underneath and thats why we had a comment in `set_exception` about not touching `*this`. This just ensures we pass a copy in so `thread_on_exception` keeps the pointer alive. I felt its just safe doing this instead of relying on developers to avoid touching `*this` after setError is called.

pritamdamania87 commented on torch/csrc/autograd/functions/accumulate_grad.h:
Looks like const ref doesn't work here (I'll keep it as a non-const ref):

```
caffe2/torch/csrc/autograd/functions/accumulate_grad.h:82:23: error: no viable overloaded '+='
        variable_grad += new_grad;
        ~~~~~~~~~~~~~ ^  ~~~~~~~~
caffe2/aten/src/ATen/TensorOperators.h:23:24: note: candidate function not viable: 'this' argument has type 'const at::Tensor', but method is not marked const
inline Tensor& Tensor::operator+=(const Tensor & other) {
                       ^
caffe2/aten/src/ATen/TensorOperators.h:26:24: note: candidate function not viable: 'this' argument has type 'const at::Tensor', but method is not marked const
inline Tensor& Tensor::operator+=(Scalar other) {
```

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."

engine.**
engine.**
engine.**

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]

pritamdamania committed:
Update on "Unify gradient accumulation between distributed autograd and local autograd
engine."

engine.**
engine.**
engine.**
engine.**
engine.**

Distributed autograd had some custom logic in terms of how we
accumulated gradients. This was mostly done early on to enable basic
functionality. Although, in the long term we should merge this logic with what
we have in the local autograd engine. A lot of work has gone into ensuring we
accumulate grads correctly and efficiently and we should reuse that as a
starting point.

We can investigate if we need further custom logic for distributed autograd
later on if we need additional optimizations.

In this PR I've merged the gradient accumulation logic and also the gradient
hooks. As a result, now gradient hooks are called in distributed autograd as
well.

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284/)

Differential Revision: [D19843284](https://our.internmc.facebook.com/intern/diff/D19843284)

[ghstack-poisoned]
